[9.x] Artisan `about` Command

jbrooksuk opened the pull request:
This PR introduces a new `about` command that displays output about the system environment and configuration of the user's application. The command is inspired by Rails' own [`about`](https://guides.rubyonrails.org/v2.3/command_line.html#about) command, though it's beautiful and has some additional powers 🪄 

![CleanShot 2022-07-13 at 12 08 46@2x](https://user-images.githubusercontent.com/246103/178720084-721affee-0a80-4d08-8aee-87c7912cb59a.png)

The command includes its own API so that applications and packages may push data into existing sections or create a new section with its own data:

```php
use Illuminate\Foundation\Console\AboutCommand;

AboutCommand::add('Laravel Vapor', 'Vapor Runtime', '1.0.0');
```

You can also add multiple items:

```php
AboutCommand::add('Environment', [
    'Laravel Cashier' => 'Stripe',
    'Laravel Nova' => '4.0.0'
]);
```

It's expected that packages will start to add their own sections of information, which could make this command output a lot of information. You can choose to filter down certain sections with `--only` and pass a comma-separated list of sections in `snake_case`.

![CleanShot 2022-07-13 at 12 10 13@2x](https://user-images.githubusercontent.com/246103/178720273-a5a1f8fe-ed95-4030-84bf-fda845035806.png)

Finally, you can choose to output this information as JSON by using the `--json` option (which can also be combined with `--only`):

![CleanShot 2022-07-13 at 12 11 40@2x](https://user-images.githubusercontent.com/246103/178720538-3df86f13-9d43-4e53-8624-aa63497a746f.png)

nunomaduro commented on src/Illuminate/Foundation/Console/AboutCommand.php:
```diff
@@ -0,0 +1,154 @@
+     *
+     * @var string
+     */
+    protected $description = 'About your application\'s environment';
```
It seems a little bit more than the Application's environment at this point?

nunomaduro commented on src/Illuminate/Support/Composer.php:
```diff
@@ -107,4 +107,21 @@ public function setWorkingPath($path)
+     */
+    public function getVersion()
+    {
+        $command = array_merge($this->findComposer(), ['-V']);
```
@jbrooksuk Did you manage to test this with different composer versions?

danharrin commented:
This is great, for package maintainers, it will allow us to more easily get debugging information from our users, which will help us fix issues. Not just version numbers, but checking if they have fallen into common traps like not linking their public storage directory.

Very nice!

sebdesign commented on src/Illuminate/Foundation/Console/AboutCommand.php:
```diff
@@ -0,0 +1,154 @@
+    protected function gatherEnvironmentData()
+    {
+        static::add('Environment', [
+            'Laravel Version' => app()->version(),
```
Why not `$this->laravel->version()`?

jbrooksuk replied:
I've switched to using this `$this->laravel` for consistency.

pxlrbt commented:
For sites with many packages, I think it might be useful to filter results. For example `php artisan about --only=environment,laravel_nova`

voidgraphics commented:
Great PR, it's a little hard to read though with the values being so far from the keys. Especially in full screen.

fgaroby commented:
@pxlrbt: wouldn't it be better using the `vendor/package_name` syntax ?

jbrooksuk commented:
Okay, I've added `--only` which accepts a command-separated list of section names in snake case, i.e. `Laravel Nova => laravel_nova`.

I've also added an `about` method to `ServiceProvider` as a shortcut to adding sections.

jbrooksuk commented:
> Great PR, it's a little hard to read though with the values being so far from the keys. Especially in full screen.

The about is dynamic based on the width of your terminal:

![CleanShot 2022-07-13 at 10 56 40@2x](https://user-images.githubusercontent.com/246103/178706579-cee99176-4b29-4cde-b17b-b17b140827c7.png)

jbrooksuk commented:
We'll address this in the Console UI Improvements PR, on which this PR is based on.

dennisprudlo commented:
@jbrooksuk Just a thought: How about getting rid of the `--only` option and use an optional argument list instead?
So `about {sections?*} {--json}`. This way the command reads better:
```shell
php artisan about environment
php artisan about environment laravel_nova some_other_section --json
```

jbrooksuk committed:
Implement `about` command

omerbaflah commented:
Great work!  @jbrooksuk  👏🏻
